src/dbnode/encoding/iterators.go

```diff
@@ -38,21 +39,71 @@ var (
 // iterators is a collection of iterators, and allows for reading in order values
 // from the underlying iterators that are separately in order themselves.
 type iterators struct {
-	values      []Iterator
-	earliest    Iterator
-	earliestIdx int
-	earliestAt  time.Time
-	filtering   bool
-	filterStart time.Time
-	filterEnd   time.Time
+	values             []Iterator
+	earliest           []Iterator
+	earliestAt         time.Time
+	filterStart        time.Time
+	filterEnd          time.Time
+	filtering          bool
+	equalTimesStrategy IterateEqualTimestampStrategy
+
+	// Used for caching reuse of value frequency lookup
+	valueFrequencies map[float64]int
 }
 
 func (i *iterators) len() int {
 	return len(i.values)
 }
 
 func (i *iterators) current() (ts.Datapoint, xtime.Unit, ts.Annotation) {
-	return i.earliest.Current()
+	numIters := len(i.earliest)
```

Comment:
could this function be implemented using a heap instead? wondering if we'd avoid sorting for each call to `current()` that way

Reply:
We can look at this in a followup change, currently none of the non-default strategies aren't likely to be called very often (only subsets of queries/requests, etc, hence using the setter at a per series iterator granularity).

When we have a more complete implementation that can actually return you the last written value, say with metadata stored alongside the value, we should then look at optimizing this.  This is mainly an ergonomic change useful for edge cases rather than supposed to be definitive and optimized.